include/zephyr/bluetooth/conn.h

```diff
@@ -1000,6 +1000,15 @@ struct bt_conn_cb {
 	 */
 	void (*disconnected)(struct bt_conn *conn, uint8_t reason);
 
+	/** @brief A connection object has been returned to the pool.
+	 *
+	 * This callback notifies the application that it might be able to
+	 * allocate a connection object. No guarantee, first come, first serve.
+	 *
+	 * Use this to e.g. re-start connectable advertising or scanning.
```

Comment:
Can you add a comment saying "please treat this as an ISR (ie no blocking) and do it from another context", something along those lines?
Basically, since this is executed in `_unref` context, restarting advertising or scanning and blocking on an HCI command to be sent will be another source of deadlocks.

The users should be guided to something alike what you did in the sample & tests.

Reply:
Expanded on this in the callback description.